File: src/aleph/dtos/tutor.py
"""Tutor API request/response DTOs (AL-221, Phase 2 TDD §6).

The wire contract for the tutor surface — the conversation read, the
Tutor-check answer, and (AL-220) the streamed send. They live here, ahead of the
send endpoint, so the turn service and the rail (AL-230/231) share one
definition instead of two that drift.

Enums reuse the model ``StrEnum``s (``MessageRole``, ``MessageSource``) so the
frontend's ``lib/api.ts`` fixes one set of strings — same word, same meaning,
prose to schema (CONTEXT.md). All addressing is by UUID; DTOs stay separate from
the ORM models (CLAUDE.md).

**A deliberate asymmetry with the Quick check** (TDD §6). ``QuickCheckDTO``
hides ``correct_index``/``explanation`` until an Attempt exists — that is Phase
1's answer-hiding invariant (W6). :class:`TutorCheckDTO` **carries** both on
delivery, by design: a Tutor check is the tutor's own **non-scoring** question,
its feedback is immediate and client-side, and nothing downstream grades it. The
invariant being protected is the *Quick check's* answer, and that protection is
behavioural (TDD D7 — prompt rule, deterministic pre-filter, W13), not a
property of this DTO.
"""

from datetime import datetime
from typing import Annotated
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, StringConstraints

from aleph.models import MessageRole, MessageSource

# One learner message (CONTEXT.md: *Turn*). Stripped of surrounding whitespace,
# required non-empty, and bounded at 2000 characters (TDD §5.5) so a pathological
# payload cannot reach the model or the DB Text column unchecked. A violation is
# a ``422`` through the shared validation envelope (``app.py``), never a 500 —
# and it is rejected *before* the stream opens, so it is an ordinary JSON error.
TutorMessageStr = Annotated[
    str, StringConstraints(strip_whitespace=True, min_length=1, max_length=2000)
]


class TutorCheckDTO(BaseModel):
    """A Tutor check as it rides on a tutor message (PRD §5.5).

    The full payload the ``pose_tutor_check`` tool produced, plus
    ``answered_index`` — the option the learner picked, ``None`` until the
    check-answer endpoint records one. Carrying ``correct_index`` and
    ``explanation`` is deliberate (see the module docstring): the card reveals
    feedback client-side the instant an option is tapped, and re-opening the
    thread has to render that same revealed state without a second round trip.

    A Tutor check is non-scoring and outside progress: answering one creates no
    Attempt and changes no lesson, progression, or §7 Attempt-derived metric
    (W12).
    """

    stem: str
    options: list[str]
    correct_index: int
    explanation: str
    answered_index: int | None = None


class MessageDTO(BaseModel):
    """One message in the thread — learner or tutor (§6).

    ``lesson_id`` + ``lesson_title`` are **the lesson the message was asked in**
    (PRD §5.8), which is what lets a revisited thread show where each turn
    happened and what 2B renders its lesson dividers from. The conversation is
    per *path*, not per lesson, so these vary down a single thread.

    ``tutor_check`` is non-null only on a tutor message that posed one. The
    learner-row ``source`` (typed vs suggestion) is deliberately **not** on the
    wire: it is the §7 entry-mix datum, carried by the product events, and the
    rail renders a learner bubble the same either way.
    """

    id: UUID
    role: MessageRole
    content: str
    lesson_id: UUID
    lesson_title: str
    tutor_check: TutorCheckDTO | None
    created_at: datetime


class ConversationResponse(BaseModel):
    """``GET /api/v1/paths/{id}/conversation`` body — the whole thread.

    Object-wrapped (never a bare array) like every other list payload, so the
    response can gain fields without a breaking change. Messages are in
    ``position`` order, oldest first.

    A path with no conversation yet answers ``200`` with an empty list, not
    ``404``: the row is created lazily on the first completed turn (TDD §4), so
    "no thread yet" is the opening state of every path rather than a missing
    resource. **Unpaginated this phase** — an accepted risk (TDD §14): a long
    thread is a whole-payload read.
    """

    messages: list[MessageDTO]


class TutorCheckAnswerRequest(BaseModel):
    """``POST /api/v1/messages/{id}/tutor-check-answer`` body.

    ``selected_index`` is the 0-based index into the check's ``options``. Unlike
    the Quick check's :class:`~aleph.dtos.lessons.AttemptRequest` — where an
    out-of-range index simply grades ``incorrect`` — this one is bounded at the
    route (``422``): nothing grades a Tutor check, so the index is *only* ever
    used to index ``options`` when re-rendering the revealed card. An
    unindexable value would be stored happily and break that render later.
    """

    selected_index: int = Field(ge=0)


class SendMessageRequest(BaseModel):
    """``POST /api/v1/paths/{id}/conversation/messages`` body (AL-220, §5.4).

    Defined here with the rest of the tutor contract so the rail and the turn
    service share one shape; the endpoint itself lands with AL-220.

    ``lesson_id`` is the lesson the question is asked in — the scope the tutor
    grounds on this phase, and the tag persisted on both rows of the turn.
    ``source`` records how the learner entered it (a suggestion button sends as
    if typed, with ``source=suggestion``), the §7 entry-mix datum.

    ``model`` is the **admin** per-message model override (TDD §5.3): a bare
    OpenRouter id from the ``MODEL_ALLOWLIST`` the session endpoint exposes to
    admins, enforced server-side exactly like Phase 1's picker (``403``
    non-admin, ``422`` off-allowlist) and resolved per request, **persisted
    nowhere** — a tutor reply is request-scoped, so there is nothing to resume
    and no column to add. The field being present is never itself a grant.
    """

    # ``model`` starts with the ``model_`` prefix pydantic protects by default;
    # the picker wire contract fixes the name (parity with the ``MODEL_TUTOR``
    # config slot and Phase 1's ``model_outline``/``model_lesson``), so opt out
    # of the protected namespace rather than rename.
    model_config = ConfigDict(protected_namespaces=())

    lesson_id: UUID
    content: TutorMessageStr
    source: MessageSource = MessageSource.TYPED
    model: str | None = None

File: src/aleph/web/frontend/src/lib/tutor.ts
// The tutor's non-streaming wire seam (AL-230, docs/api.md ## Tutor): the
// conversation read + clear routes, their types, and the one query identity the
// rail reads. Everything here goes through `apiFetch`/`apiV1Path` exactly like
// `lib/api.ts`'s Phase 1 calls — only the *streamed* send lives apart, in
// `lib/tutor-stream.ts`, and only because it must read the body progressively.
//
// **The thread is ordinary query state, not a poll target** (TDD §8). Phase 1's
// generation surfaces poll because work outlives the request; a tutor reply is
// request-scoped, arrives on its own stream, and is appended to this cache when
// it settles. Nothing here carries a `refetchInterval`, and a *failed* reply
// invalidates nothing — a turn exists whole or not at all (D2), so there is
// never server state the client is out of step with.

import { queryOptions, skipToken } from "@tanstack/react-query";
import { apiFetch, apiV1Path } from "./api";
import type { TutorCheck } from "./tutor-stream";

/** Who spoke (docs/api.md). Learner and tutor messages pair into a **turn**. */
export type TutorRole = "learner" | "tutor";

/**
 * One message in the thread. There is **one conversation per path**, so
 * `lesson_id`/`lesson_title` — the lesson the message was asked in — vary down a
 * single thread (PRD §5.8; Phase 2B renders them as dividers).
 *
 * Note the payload carries no `position`: the array order **is** the order.
 */
export interface ConversationMessage {
  id: string;
  role: TutorRole;
  content: string;
  lesson_id: string;
  lesson_title: string;
  /** Set only on a tutor message that posed a Tutor check (AL-231 renders it). */
  tutor_check: TutorCheck | null;
  created_at: string;
}

/** `GET /api/v1/paths/{id}/conversation` — object-wrapped, oldest first. */
export interface Conversation {
  messages: ConversationMessage[];
}

/** The whole thread for a path. `200 {messages: []}` when none exists yet. */
export function getConversation(pathId: string): Promise<Conversation> {
  return apiFetch<Conversation>(apiV1Path(`/paths/${pathId}/conversation`));
}

/**
 * **New conversation** (PRD §5.8): drop the thread for this path. `204` and
 * idempotent; cascades to its messages; touches no Phase 1 state. Destructive
 * and not undoable — the caller MUST confirm first, like `deletePath`.
 */
export function clearConversation(pathId: string): Promise<void> {
  return apiFetch<void>(apiV1Path(`/paths/${pathId}/conversation`), { method: "DELETE" });
}

/**
 * Record the learner's Tutor-check choice (AL-231, route from AL-221) → `204`.
 *
 * Addressed by **message** id: a check rides the tutor message that posed it.
 * `selected_index` indexes that check's `options` — out of range is a `422`,
 * because nothing grades a Tutor check and the index is only ever used to index
 * `options` when re-rendering the revealed card.
 *
 * **This call grades nothing.** The delivered payload already carries
 * `correct_index` + `explanation` (TDD §6), so the card reveals from what it is
 * already holding and calls this *afterwards*, only so a revisited thread
 * renders revealed. Re-answering overwrites (last-wins) — deliberately unlike
 * the Quick check's first-wins Attempt, because an Attempt is graded and feeds
 * the §7 metrics, and neither is true here.
 */
export function answerTutorCheck(messageId: string, selectedIndex: number): Promise<void> {
  return apiFetch<void>(apiV1Path(`/messages/${messageId}/tutor-check-answer`), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ selected_index: selectedIndex }),
  });
}

/**
 * TanStack query key for one path's thread. The `"tutor"` head is its own
 * namespace rather than a branch of `["paths", …]`: lesson completion
 * invalidates that whole prefix (`PATHS_QUERY_PREFIX`), and a conversation has
 * no business being refetched because a lesson was marked complete.
 */
export function conversationQueryKey(pathId: string): readonly ["tutor", "conversation", string] {
  return ["tutor", "conversation", pathId] as const;
}

/**
 * THE conversation query — key + fetcher paired in one place (the house rule
 * from `sessionQueryOptions`). Pass `null` when the rail's entry point is not
 * rendered (flag off, or a lesson with no generated content): the query idles on
 * `skipToken` and the gated surface costs no request at all.
 */
export function conversationQueryOptions(pathId: string | null) {
  return queryOptions({
    queryKey: conversationQueryKey(pathId ?? "idle"),
    queryFn: pathId === null ? skipToken : () => getConversation(pathId),
  });
}
